### PR TITLE
Organise testing utils

### LINKED ISSUE #564: Organise testing fixtures and utilities so that they can be made use of by application code
- mock devices should be provided next to real devices
- there should be a `dodal.testing_utils` for generic stuff
- mock devices could be rexported from `dodal.testing_utils` as fixtures

### PR DESCRIPTION
**Pull request by dperl-dls**
Fixes #564

Moves many mock device test fixtures and utility functions to a central place under `src/`, so that they can easily be reused by application code

### Instructions to reviewer on how to test:
1. See that no tests are broken

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)
- [ ] If changing the API for a pre-existing device, ensure that any beamlines using this device have updated their Bluesky plans accordingly

### CODE REVIEW

File: src/dodal/testing_utils/device_fixtures.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        TEST_APERTURE_POSITIONS
+    )
+    return aperture_positions
+
+
+@pytest.fixture
+async def mock_aperturescatterguard_with_call_log():
+    call_log = MagicMock()
+    ap_sg = await get_mock_ap_sg()
+    with (
```
Comment (DominicOram) [inline src/dodal/testing_utils/device_fixtures.py:53]: Should: `patch_ophyd_async_motor` isn't a context manager

File: src/dodal/testing_utils/device_fixtures.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        patch_ophyd_async_motor(ap_sg.aperture.x, call_log=call_log),
+        patch_ophyd_async_motor(ap_sg.aperture.y, call_log=call_log),
+        patch_ophyd_async_motor(ap_sg.aperture.z, call_log=call_log),
+        patch_ophyd_async_motor(ap_sg.scatterguard.x, call_log=call_log),
+        patch_ophyd_async_motor(ap_sg.scatterguard.y, call_log=call_log),
```
Comment (DominicOram) [inline src/dodal/testing_utils/device_fixtures.py:58]: Should: Can we do this patching in `get_mock_ap_sg`?

File: src/dodal/testing_utils/device_fixtures.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    return make_fake_device(DualBacklight)(name=f"backlight: {request.node.name}")
+
+
+@pytest.fixture
+def mock_eiger(request: pytest.FixtureRequest):
+    yield get_mock_eiger(create_new_detector_params(), request.node.name)
+
+
+@pytest.fixture
+def mock_eiger_and_stage(mock_eiger):
```
Comment (DominicOram) [inline src/dodal/testing_utils/device_fixtures.py:111]: Nit: Rename to something like `mock_eiger_with_noop_stage`?

**Review by DominicOram — CHANGES_REQUESTED** [review]
Thank you! I'm not convinced about the structure as whole yet. Could you make the changes in hyperion so we have an example of how this improves code in a downstream app? Then I think maybe it needs a brief in person discussion

File: src/dodal/testing_utils/device_fixtures.py (hunk 1 above)
Comment (dperl-dls) [reply src/dodal/testing_utils/device_fixtures.py:53]: yes it is, it's type-hinted wrong in ophyd-async, fix has been merged but not released, doesn't affect functionality

File: src/dodal/testing_utils/device_fixtures.py (hunk 2 above)
Comment (dperl-dls) [reply src/dodal/testing_utils/device_fixtures.py:58]: no, it leads to circular import errors if we import test utils to device classes and then device classes to test utils

**Comment by dperl-dls** [comment]
never mind, this is too hard, I now think it's better to just copy and paste stuff everywhere

**Pull request closed without merging**